lib/cstruct.ml

```diff
@@ -128,7 +128,8 @@ let debug t =
 let sub t off0 len =
   let off = t.off + off0 in
   let offlen = off + len in
-  if off < off0 || offlen < off || off0 < 0 || len < 0 || not (check_bounds t (offlen)) then err_sub t off0 len
+  let old_end = t.off + t.len in
+  if offlen > old_end || off < off0 || offlen < off || off0 < 0 || len < 0 || not (check_bounds t (offlen)) then err_sub t off0 len
```

Comment:
Can we simplify this check? If it's a sub-cstruct of the input, it must be within the bigarray too.

Perhaps something like:

```ocaml
let sub t off0 len =
  let new_start = t.off + off0 in
  let new_end = new_start + len in
  let old_end = t.off + t.len in
  if new_start <= new_end && new_start >= t.off && new_end <= old_end then { t with off = new_start; len }
  else err_sub t off0 len
```

Reply:
maybe, I'd leave this as a potential future PR.

Reply:
I think it should have a TODO at least, to avoid confusing future readers.

Reply:
@hannesm the one above is easier to read -- why not just use that one instead, since we've got to test a bunch of uses of `sub` anyway with whichever one we go with

Reply:
It's good to be careful here. For speed, we don't do any further bounds checking after creating a cstruct, so we need to get this right. Here is my argument for the correctness of the function above.

`Cstruct.sub` should select what a programmer intuitively expects a sub-cstruct to be. I imagine holding out my hands, with the left representing the start offset and the right the end. I think of a sub-cstruct as any span within this range. If I move my left hand only to the right (`new_start >= t.off`), and my right hand only to the left `new_end <= old_end`, and they don't cross `new_start <= new_end`, then I feel sure the result will be a valid sub-cstruct. And if I violate any one of these constraints (e.g. moving my left hand further left), then I feel sure that the result wouldn't be something I'd consider to be a sub-cstruct.

Therefore, if this code returns a new cstruct then `new_start` and `new_end` will be in range, and I think it's clear enough that they will also be what the user expected.

However, we don't store `new_end`; we instead store `len`, from which `new_end` is calculated. Is it possible to provide an out-of-range `len` that still results in an in-range `new_end`, given that we are using modular arithmetic? No. For any valid `new_end` there must be some correct `len`, and `new_end = new_start + len` has only one solution for `len` given `new_end` and `new_start`.

(The manual says `All operations are taken modulo 2^Sys.int_size` so we don't have to worry about crazy C-style Undefined Behaviour). 

If the above isn't convincing, there is another way to see that modular arithmetic doesn't cause a problem: if the function returns a new cstruct then none of the three addition operations in the function can have wrapped.

Note that if `x` is non-negative, and `x + y` wraps, then `x + y` must be negative. This is easy to see with modular arithmetic because if `y` is negative then the two arguments will cancel to some degree the result cannot be further from zero than one of the arguments. If `y` is positive then `x + y` can wrap, but even `max_int + max_int` doesn't wrap all the way to zero.

The three possibly-wrapping operations are:

`new_start = t.off + off0`. `t.off` is non-negative so if this wraps then `new_start` will be negative and will fail the `new_start >= t.off` test.

`new_end = new_start + len`. The above test ensures that `new_start` is non-negative in any successful return. So if this wraps then `new_end` will be negative and will fail the `new_start <= new_end` test.

`old_end = t.off + t.len`. This uses only the existing trusted values. It could only wrap if the underlying bigarray had a negative length!

It would be good if someone could check the reasoning above and/or provide a simpler argument that the alternative code is correct.